src/rust_api.rs

```diff
@@ -0,0 +1,1561 @@
+// Copyright 2016 6WIND S.A. <quentin.monnet@6wind.com>
+//
+// Licensed under the Apache License, Version 2.0 <http://www.apache.org/licenses/LICENSE-2.0> or
+// the MIT license <http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+//! Module provides API to create eBPF programs by Rust programming language
+
+use ebpf::Insn;
+
+/// Represents single eBPF instruction
+pub trait Instruction: Sized {
+    /// returns instruction opt code
+    fn opt_code_byte(&self) -> u8;
+
+    /// returns destination register
+    fn dst(&self) -> u8;
+
+    /// returns source register
+    fn src(&self) -> u8;
+
+    /// returns offset bytes
+    fn offset(&self) -> i16;
+
+    /// returns immediate value
+    fn imm(&self) -> i32;
+
+    /// sets destination register
+    fn dst_reg(self, dst: u8) -> Self;
+
+    /// sets source register
+    fn src_reg(self, src: u8) -> Self;
+
+    /// sets offset bytes
+    fn offset_bytes(self, offset: i16) -> Self;
+
+    /// sets immediate value
+    fn immediate(self, imm: i32) -> Self;
+}
+
+pub trait IntoBytes {
+    type Bytes;
+
+    fn into_bytes(self) -> Self::Bytes;
+}
+
+impl<'i, I: Instruction> IntoBytes for &'i I {
+    type Bytes = Vec<u8>;
+
+    fn into_bytes(self) -> Self::Bytes {
+        let mut buffer = Vec::with_capacity(8);
+        buffer.push(self.opt_code_byte());
+        buffer.push(self.src() << 4 | self.dst());
+        buffer.push((self.offset() & 0x00_ff) as u8);
+        buffer.push(((self.offset() & 0xff_00) >> 8) as u8);
+        buffer.push((self.imm() & 0x00_00_00_ff) as u8);
+        buffer.push(((self.imm() & 0x00_00_ff_00) >> 8) as u8);
+        buffer.push(((self.imm() & 0x00_ff_00_00) >> 16) as u8);
+        buffer.push(((self.imm() & 0xff_00_00_00) >> 24) as u8);
+        buffer
+    }
+}
+
+/// BPF instruction stack in byte representation
+pub struct BpfCode {
+    instructions: Vec<u8>
+}
+
+impl BpfCode {
+
+    pub fn new() -> Self {
+        BpfCode { instructions: vec![] }
+    }
+
+    pub fn mov_add(&mut self, source: Source, arch: Arch) -> Move {
```

Comment:
Again, I'm not sure the `mov_` prefix will help end users understanding what the function does.